src/elm/external.ts

```diff
@@ -22,21 +25,49 @@ export class Retrieve extends Expression {
   }
 
   exec(ctx: Context) {
-    let records = ctx.findRecords(this.templateId != null ? this.templateId : this.datatype, this);
-    let codes = this.codes;
-    if (this.codes && typeof this.codes.exec === 'function') {
-      codes = this.codes.execute(ctx);
-      if (codes == null) {
+    // Object with retrieve information to pass back to patient source
+    const retrieveDetails: RetrieveDetails = {
+      datatype: this.datatype
+    };
+
+    let resolvedCodes: Code[] | ValueSet | undefined;
+    if (this.codes) {
+      resolvedCodes = this.codes.execute(ctx);
+
+      if (resolvedCodes == null) {
         return [];
       }
-    }
-    if (codes) {
-      records = records.filter((r: any) => this.recordMatchesCodesOrVS(r, codes));
+
+      retrieveDetails.codes = resolvedCodes;
     }
     // TODO: Added @dateProperty check due to previous fix in cql4browsers in cql_qdm_patient_api hash: ddbc57
+    let range: Interval | undefined;
     if (this.dateRange && this.dateProperty) {
-      const range = this.dateRange.execute(ctx);
-      records = records.filter((r: any) => range.includes(r.getDateOrInterval(this.dateProperty)));
+      range = this.dateRange.execute(ctx);
+
+      retrieveDetails.dateRange = range;
+      retrieveDetails.dateProperty = this.dateProperty;
+    }
+
+    if (this.templateId) {
+      retrieveDetails.templateId = this.templateId;
+    }
+
+    if (resolvedCodes && this.codeProperty) {
+      retrieveDetails.codeProperty = this.codeProperty;
+    }
+
+    let records = ctx.findRecords(
+      this.templateId != null ? this.templateId : this.datatype,
+      retrieveDetails
+    );
+
+    if (resolvedCodes) {
+      records = records.filter((r: any) => this.recordMatchesCodesOrVS(r, resolvedCodes));
+    }
+
+    if (range && this.dateProperty) {
+      records = records.filter((r: any) => range?.includes(r.getDateOrInterval(this.dateProperty)));
```

Comment:
Since you did a null-check before coming into this block, I don't think the `?` guard is necessary.

Reply:
You would think!!! Check this out:

Before the scope of the `if`, range is `Interval | undefined`. Logical.

<img width="439" alt="image" src="https://user-images.githubusercontent.com/16297930/157463269-9bf0aaf2-ed74-4e5f-86b9-dea49e52a947.png">

Within the scope of the `if`, it's STILL `Interval | undefined`

<img width="975" alt="image" src="https://user-images.githubusercontent.com/16297930/157463350-213f2d64-0c6c-47db-8b82-db01076c6c9d.png">


BUT within the scope of the `if` and outside the anonymous function within `filter`, it's just `Interval`:

<img width="376" alt="image" src="https://user-images.githubusercontent.com/16297930/157463450-245c0de0-b1ac-4662-ba56-1a2e23468c08.png">


Legit no idea